src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Shell/Helpers/ShellListPageHelpers.cs

```diff
@@ -0,0 +1,117 @@
+using Microsoft.CmdPal.Extensions.Helpers;
+
+namespace Microsoft.CmdPal.Ext.Shell.Helpers;
+
+public class ShellListPageHelpers
+{
+    private static readonly CompositeFormat WoxPluginCmdCmdHasBeenExecutedTimes = System.Text.CompositeFormat.Parse(Properties.Resources.wox_plugin_cmd_cmd_has_been_executed_times);
+    private readonly SettingsManager _settings;
+
+    public ShellListPageHelpers(SettingsManager settings)
+    {
+        _settings = settings;
+    }
+
+    private ListItem GetCurrentCmd(string cmd)
+    {
+        ListItem result = new ListItem(new ExecuteItem(cmd, _settings))
+        {
+            Title = cmd,
+            Subtitle = Properties.Resources.wox_plugin_cmd_plugin_name + ": " + Properties.Resources.wox_plugin_cmd_execute_through_shell,
+            Icon = new(string.Empty),
+        };
+
+        return result;
+    }
+
+    private List<ListItem> GetHistoryCmds(string cmd, ListItem result)
+    {
+        IEnumerable<ListItem> history = _settings.Count.Where(o => o.Key.Contains(cmd, StringComparison.CurrentCultureIgnoreCase))
+            .OrderByDescending(o => o.Value)
+            .Select(m =>
+            {
+                if (m.Key == cmd)
+                {
+                    // Using CurrentCulture since this is user facing
+                    result.Subtitle = Properties.Resources.wox_plugin_cmd_plugin_name + ": " + string.Format(CultureInfo.CurrentCulture, WoxPluginCmdCmdHasBeenExecutedTimes, m.Value);
+                    return null;
+                }
+
+                var ret = new ListItem(new ExecuteItem(m.Key, _settings))
+                {
+                    Title = m.Key,
+
+                    // Using CurrentCulture since this is user facing
+                    Subtitle = Properties.Resources.wox_plugin_cmd_plugin_name + ": " + string.Format(CultureInfo.CurrentCulture, WoxPluginCmdCmdHasBeenExecutedTimes, m.Value),
+                    Icon = new(string.Empty),
+                };
+                return ret;
+            }).Where(o => o != null).Take(4);
+        return history.ToList();
+    }
+
+    public List<ListItem> Query(string query)
+    {
+        ArgumentNullException.ThrowIfNull(query);
+
+        List<ListItem> results = new List<ListItem>();
+        var cmd = query;
+        if (string.IsNullOrEmpty(cmd))
+        {
+            results = ResultsFromlHistory();
+        }
+        else
+        {
+            var queryCmd = GetCurrentCmd(cmd);
+            results.Add(queryCmd);
+            var history = GetHistoryCmds(cmd, queryCmd);
+            results.AddRange(history);
+        }
+
+        for (var i = 0; i < results.Count; i++)
+        {
+            var currItem = results[i];
+            currItem.MoreCommands = LoadContextMenus(currItem).ToArray();
+        }
+
+        return results;
+    }
+
+    public List<CommandContextItem> LoadContextMenus(ListItem listItem)
+    {
+        var resultlist = new List<CommandContextItem>
```

Comment:
Future thought: we stick the other kinds of run options into the context menu. So the default is whatever's in the settings, and the context menu has "run in powershell", etc.

Reply:
(Trying to remember how the settings for each extension work as well, but this gave me a thought; if they're not already as a future idea, could be neat to have a quick link to settings when the context of the app changes to a specific extension.)

Reply:
Yeah so there is https://github.com/zadjii-msft/PowerToys/issues/122. 

I think that what you're saying @michael-hawker makes sense here though... maybe we can have that show up easily for all extensions if you are in an extension experience. I like it. 

I'll create an issue for quick link to settings when the context of the app changes to a specific extension + the additional run options into the context menu.

Reply:
#193 to track the additional options in the context menu

#195 for tracking linking to settings in some consistent way when the app changes to a specific extension

Reply:
Not going to handle this feature in this PR - but as mentioned in previous comment, I've added an issue for it that we can track.